### ISSUE TITLE
Don't say that a `yield` statement in an annotation scope in a function is "not in a function"

### ISSUE THREAD

**Issue by ntBre** [astral-sh/ty#299]
I think there is still something here. With 
```py
def _():
# error: [invalid-type-form] "`yield` expressions are not allowed in type expressions"
# error: [invalid-syntax] "yield expression cannot be used within a TypeVar bound"
    type X[T: (yield 1)] = int

def _():
# error: [invalid-type-form] "`yield` expressions are not allowed in type expressions"
# error: [invalid-syntax] "yield expression cannot be used within a type alias"
    type Y = (yield 1)
```

I still get the outside of a function errors

```bash
semantic_syntax_errors.md - Semantic syntax error diagnostics - Invalid expression

  crates/ty_python_semantic/resources/mdtest/diagnostics/semantic_syntax_errors.md:304 unexpected error: 16 [invalid-syntax] "`yield` statement outside of a function"
  crates/ty_python_semantic/resources/mdtest/diagnostics/semantic_syntax_errors.md:309 unexpected error: 15 [invalid-syntax] "`yield` statement outside of a function"
```

_Originally posted by @maxmynter in https://github.com/astral-sh/ruff/pull/17748#discussion_r2082080479_

**Comment by ntBre** [comment]
In case someone wants to get to this before me, I think you'll just need to update [this function](https://github.com/astral-sh/ruff/blob/ec2fdadb35e618624d865e7a241452d6aff833e4/crates/ty_python_semantic/src/semantic_index/builder.rs#L2503-L2506) to look more like `in_await_allowed_context` a few lines above it. And that's called [here](https://github.com/astral-sh/ruff/blob/ec2fdadb35e618624d865e7a241452d6aff833e4/crates/ruff_python_parser/src/semantic_errors.rs#L765) in case it turns out to be more involved than that.

**Comment by maxmynter** [comment]
Having stumbled upon this, I would love to take a jab at it :)

**Comment by maxmynter** [comment]
PR ready in the Ruff repository: https://github.com/astral-sh/ruff/pull/18008
